### PR TITLE
(QENG-3919) spike for implementation of all-or-nothing checkout

### PR DESCRIPTION
**Pull request by rick**
:warning: do not merge :warning:

Just getting this into the system so we can look at it and discuss whether this would work, etc.

cc @joshcooper @heathseals

### CODE REVIEW

File: lib/vmpooler/api/v1.rb [hunk 1]
```diff
@@ -335,40 +339,56 @@ def checkout_vm(template, result)
...
-      end
+        failed = false
+        vms = []
 
-      if (available == 1)
-        result['ok'] = true
-
-        jdata.each do |key, val|
+        jdata.each do |template, count|
           val.to_i.times do |_i|
```
Comment (rick) [inline lib/vmpooler/api/v1.rb]: s/val/count/

File: spec/vmpooler/api/v1_spec.rb [hunk 2]
```diff
@@ -265,6 +265,32 @@ def app()
         expect_json(ok = true, http = 200)
       end
 
+      it 'fails when not all requested vms can be allocated' do
+        allow(redis).to receive(:spop).with('vmpooler__ready__pool1').and_return 'abcdefghijklmnop'
```
Comment (rick) [inline spec/vmpooler/api/v1_spec.rb]: Don't really like these very mockist tests, but this is the shortest path from the current suite.

If we have a redis available, seems like we should set up state, perform operations, validate state, and then teardown state.

File: spec/vmpooler/api/v1_spec.rb [hunk 3]
```diff
@@ -265,6 +265,32 @@ def app()
...
+        expected = { ok: false }
+
+        expect(last_response.body).to eq(JSON.pretty_generate(expected))
+        expect_json(ok = false, http = 200) # which HTTP status code?
+      end
+
+      it 'returns any checked out vms when not all requested vms can be allocated' do
+        allow(redis).to receive(:spop).with('vmpooler__ready__pool1').and_return 'abcdefghijklmnop'
+        allow(redis).to receive(:spop).with('vmpooler__ready__pool2').and_return nil
+        expect(redis).to receive(:spush).with("vmpooler__ready__pool1", "abcdefghijklmnop")
```
Comment (joshcooper) [inline spec/vmpooler/api/v1_spec.rb]: Do you want to test for the case where it needs to return multiple instances of the same template, and multiple instances from different templates?

**Comment by joshcooper** [comment]
Just have the one question about testing N boundary cases. Otherwise looks good to me!

File: spec/vmpooler/api/v1_spec.rb (hunk 2 above)
Comment (rick) [inline spec/vmpooler/api/v1_spec.rb]: cc #152

File: spec/vmpooler/api/v1_spec.rb (hunk 3 above)
Comment (rick) [inline spec/vmpooler/api/v1_spec.rb]: Very much :+1: on that. I'm headed out for the day, but feel free to add those if you want.

**Comment by rick** [comment]
Cool, I'll add those tests soon. We'll want to get this onto the ci.next vmpooler instance and test drive it. This will eventually need to go out to production, wondering if we have any techniques for minimizing risk, in case we missed something.

If this works as desired, what will be the impact on clients using it? We are basically going to not be returning partial sets of requested vms -- we'll either return the full set or no nodes. Seems like that shouldn't be a problem, as callers would be retrying requests in the old behavior.

cc @puppetlabs/cinext

File: lib/vmpooler/api/v1.rb [hunk 4]
```diff
@@ -45,35 +45,39 @@ def alias_deref(hash)
       newhash
     end
 
-    def checkout_vm(template, result)
-      vm = backend.spop('vmpooler__ready__' + template)
+    def fetch_single_vm(template)
+      backend.spop('vmpooler__ready__' + template)
```
Comment (waynr) [inline lib/vmpooler/api/v1.rb]: Does this cause messages to be emitted to the log file? If so, I can imagine the log getting very noisy during periods of high load when there could be large numbers of failures for a given vm collection request.

Comment (rick) [inline lib/vmpooler/api/v1.rb]: If it does, it shouldn't happen any more often than before -- we effectively factored this out and changed the order of operations, but this isn't net new calls to spop.

Comment (joshcooper) [inline lib/vmpooler/api/v1.rb]: It looks like vmpooler explicitly emits log messages:

```ruby
        $logger.log('s', "[>] [#{pool}] '#{vm}' moved to 'ready' queue")
```

So I'm thinking there shouldn't be any added verbosity. Unless our redis library also logs somewhere?

**Comment by rick** [comment]
Closing in favor of #153, which is targeted to the `ci.next` branch.

**Pull request closed without merging**